### PR TITLE
[tune] Optional forcible trial cleanup, return default autofilled metrics even if Trainable doesn't report at least once

### LINKED ISSUE #18745: [Bug] Tune trials terminated with ERROR have no "loc"
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Tune

### What happened + What you expected to happen

A Tune trial that fails with ERROR has no "loc" field. This makes it very hard to find where it was running (don't know the worker IP).

### Reproduction script

Run any tune trial in a cluster, and view the table on failure.

cc @richardliaw 

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR adds two features:
1. An option to forcibly terminate trials scheduled for cleanup after a grace period of 60 seconds. By default this is False (old behavior). This is controlled by a new env var, `TUNE_FORCE_TRIAL_CLEANUP`.
2. An assurance that even if the Trainable doesn't get to report results at least once (due to an error or termination), auto-filled metrics will be returned.

## Related issue number

Closes https://github.com/ray-project/ray/issues/18745

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/trial.py [hunk 1]
```diff
@@ -394,6 +395,31 @@ def _setup_resources(self, log_always: bool = False):
...
+        # 2. If the trial has not reported at least once but we have the
+        #    future for the default results dict, (obtained through
+        #    Trainable.get_auto_filled_metrics), we get that future
+        #    and return it.
+        # 3. In the worst case where we have nothing, we just set the
+        #    trial_id and done keys and return that.
+        result = self._last_result
+        if not result and self._default_result_future:
+            if isinstance(self._default_result_future, ray.ObjectRef):
+                self._default_result_future = ray.get(
```
Comment (xwjiang2010) [inline python/ray/tune/trial.py:412]: Can we rename it to _default_result_or_future to be consistent?

File: python/ray/tune/ray_trial_executor.py [hunk 2]
```diff
@@ -132,6 +138,11 @@ def cleanup(self, partial: bool = True):
                 logger.warning(
                     "Skipping cleanup - trainable.stop did not return in "
                     "time. Consider making `stop` a faster operation.")
+                if not partial and self._force_cleanup:
+                    logger.warning(
+                        "Forcing trainable cleanup by terminating actors.")
+                    self._cleanup_map = {}
```
Comment (xwjiang2010) [inline python/ray/tune/ray_trial_executor.py:144]: Does this actually stop the actors? Will this introduce dangling actors in remote nodes?

**Comment by xwjiang2010** [comment]
Thanks for picking this up. Can you add a screenshot of what the autofilled metrics would look like?

File: python/ray/tune/trial_runner.py [hunk 3]
```diff
@@ -858,7 +858,9 @@ def _process_trial_result(self, trial, result):
             result = trial.last_result
             result.update(done=True)
 
-        self._total_time += result.get(TIME_THIS_ITER_S, 0)
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/tune/trial_runner.py:861]: `TIME_THIS_ITER_S` may now be None (if the trainable doesn't report at least once)

**Review by krfricke — COMMENTED** [review]
Looks good so far!

**Review by xwjiang2010 — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
Awesome, thanks!

**Pull request merged**